mqtt/mqtt-broker/src/broker.rs

```diff
@@ -797,6 +814,99 @@ where
 
         Ok(())
     }
+
+    async fn notify_subscription_change(&mut self, client_id: &ClientId) {
+        let topics: String = self
+            .sessions
+            .get(client_id) // find session that changed
+            .and_then(|session| match session {
+                // get session state
+                Session::Transient(connected) => Some(connected.state()),
+                Session::Persistent(connected) => Some(connected.state()),
+                Session::Offline(offline) => Some(offline.state()),
+                _ => None,
+            })
+            .map(|state| state.subscriptions().keys()) // get topic filters
+            .map(|topics| {
+                topics
+                    .map(|t| t.as_ref())
+                    .collect::<Vec<&str>>()
+                    .join(r"\u{0000}")
+            }) // join to string for payload
+            .unwrap_or_default(); // no subscriptions is default
+
+        let publication = proto::Publication {
+            topic_name: format!("$sys/subscriptions/{}", client_id),
+            qos: proto::QoS::AtMostOnce, //no ack
+            retain: true,
+            payload: Bytes::from(topics),
+        };
+
+        self.publish_all(publication).await.unwrap();
+    }
+
+    async fn notify_disconnect(&mut self, client_id: &ClientId) {
+        if let None = self.sessions.get(client_id) { // If transient session is closed, remove its subscriptions
+            let publication = proto::Publication {
+                topic_name: format!("$sys/subscriptions/{}", client_id),
+                qos: proto::QoS::AtMostOnce, //no ack
+                retain: true,
+                payload: Bytes::from("".to_owned()),
```

Comment:
nit:
```suggestion
                payload: "".into(),
```

Reply:
Is that better? I think `.to_owned()` is more clear

Reply:
In my limited understanding they both will do the same thing in this case, but `.to_owned()` has a bit different purpose in general. If you need to convert one type into another (like here &str to Bytes) I would use `.into()`. Not a big deal, just nitpicking :)

Reply:
oh, I forgot bytes impls From<&'static str>. I was using .to_owned to convert to a String, then From to turn the string into bytes. In this case it is better just to use .into()